deviseを使ったユーザー認証の実装

ham-cap opened the pull request:
「deviseを使ってユーザー認証を実装する」の課題を提出したします。
ご確認のほどよろしくお願いいたします。

ham-cap committed:
deviseをインストール・Userモデルを作成

ham-cap committed:
ユーザーの一覧表示・ログイン済みか否かの判定・ログアウト機能をそれぞれ実装

ham-cap committed:
ログインとログアウト時のリダイレクト先を変更・ユーザー一覧に表示している情報を詳細画面へのリンクに変更・ユーザー詳細画面の追加

・ログインとログアウトの際に/までリダイレクトされ、別のページに移動するにはURLを直接打ち込むしかなかったため、
それぞれ本の一覧とログイン画面に遷移するよう変更。
・ユーザー一覧に表示している情報（メールアドレス）をユーザー詳細画面へのリンクに変更し、ユーザー詳細画面のviewも一部修正。

ham-cap committed:
usersテーブルにカラムを追加・ユーザー情報更新時のリダイレクト先を変更・ログインユーザーが自分自身以外のユーザー情報を編集できないようにした

・ユーザー詳細ページでニックネーム等を表示させたいため、usersテーブルにnickname,zipcode,address,introduceカラムを追加した
・ユーザー情報更新時のリダイレクト先がデフォルトでは`/`になっているためユーザー一覧へ戻るようにした
・初期状態ではユーザー同士が自由に互いの情報を編集できてしまうため、他人の情報を編集できないようにした

ham-cap committed:
ユーザーの一覧、詳細画面にリンク等を追加・ログインパスワードの再設定機能を追加

・ユーザーの一覧画面から別のページへ移動できるリンクがなかったため本の一覧画面へのリンクを設置
・ユーザーの詳細画面にニックネーム、郵便番号、住所、自己紹介を追加
・パスワードの再設定をする際のメール送信でエラーになったためGemfileにnet-smtp、net-imap、net-popを追加してインストール
・ユーザーの一覧と詳細画面には未ログイン状態でもアクセスできてしまっていたためusers_controllerのbefore_actionにもAuthenticate_user!を追加

ham-cap committed:
deviseが作成したviewを日本語化

ham-cap committed:
ユーザー一覧をページング処理

ham-cap committed:
ユーザーの新規登録画面にニックネーム等の入力欄を追加

・新規登録時にメールとパスワードしか入力欄がなかったため、ニックネーム、郵便番号、住所、自己紹介文の欄を追加した。

ham-cap committed:
rubocopからの指摘事項を修正

ham-cap committed:
不要なファイルを削除

・自動で作成されたものの、使用していないファイルを削除した。

yoshitsugu commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,70 @@
+      render action: :edit
+    else
+      redirect_to users_path
+    end
```
そもそも `registrations_controller` のデフォルトの挙動で、ログインしているユーザー自身を `@user` にロードしているため、ここは不要かなと思うのですが、いかがでしょうか。

ham-cap replied:
コメントアウトして検証してみました！
仰るとおりここは不要でしたので、初期状態の`super`にしてコメントアウトします。

yoshitsugu commented on app/views/users/mailer/confirmation_instructions.html.erb:
```diff
@@ -0,0 +1,5 @@
+<p><%= t('.greeting', recipient: @email) %></p>
+
+<p><%= t('.instruction') %></p>
+
+<p><%= link_to t('.action'), confirmation_url(@resource, confirmation_token: @token) %></p>
```
今回使ってないmailerのtemplateは消してしまってかまいません〜

ham-cap replied:
削除します！

yoshitsugu commented on app/views/users/registrations/edit.html.erb:
```diff
@@ -0,0 +1,63 @@
+  <%= render "users/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label User.human_attribute_name(:nickname) %><br />
```
new.html.erbのほうには `<br />` ついていないんですが、なぜ edit.html.erbのほうにだけついているんでしょうか？ :eyes: 何か意図があればおしらせください〜

ham-cap replied:
edit.html.erbのほうを取り忘れました！
`<br />`を削除した理由はnew.html.erbをブラウザで表示した際にラベルと入力フォームの間に一行空いてしまうのが不自然に思えたためです。edit.html.erbのほうも揃えて削除するつもりでしたが、後回しにしているうちに失念したという次第です🙇‍♂️

yoshitsugu commented on db/migrate/20220529030658_change_datatype_zipcode_of_users.rb:
```diff
@@ -0,0 +1,9 @@
+class ChangeDatatypeZipcodeOfUsers < ActiveRecord::Migration[6.1]
+  def up
+    change_column :users, :zipcode, :text
```
こういったマージ前のPR内の修正であればマイグレーションファイルを編集してしまってかまわないです。
マージ前であればご自身の開発環境にしか影響はないため、実務においても基本その方針で問題ないかと思います(チームによっては違うかもしれません)

ham-cap replied:
では、手順としてはそもそも20220529030658_change_datatype_zipcode_of_users.rbを作成するのではなく、20220523212231_add_zipcode_address_introduce_to_users.rbを`rails db:rollback`で差し戻して、内容修正後に再度`rails db:migrate`で反映させる、という理解で大丈夫でしょうか？👀

yoshitsugu replied:
@ham-cap はい、そんな感じで大丈夫です。手元であれば一旦 `rails db:drop` して `rails db:create` でDBごと作り直しちゃってもいいと思います。

yoshitsugu reviewed (commented):
気になったところコメントしました！

ham-cap committed:
registrations_controller.rbのeditをコメントアウト・メール用の不要なテンプレートを削除・ユーザー編集画面の不要な改行を削除

・registrations_controller.rbのeditアクションでcurrent_userを確認する処理は不要だったためコメントアウト
・メール用のテンプレートのうちパスワードを忘れた場合の再設定用以外のテンプレートは使用していないため削除
・ユーザー編集画面の入力フォームとラベルの間が開きすぎていたため改行を削除

ham-cap committed:
rubocopの指摘事項を修正

yoshitsugu reviewed (commented):
あとはマイグレーションだけ直せばOKそうですね。

ham-cap committed:
マイグレーションファイルを修正

・zipcodeカラムのデータ型をintegerからtextに変更する際、新規でマイグレーションファイルを作成し、upとdownで修正していたが
マージ前のもののため、カラムを追加した際のマイグレーションファイルを修正することとした。
データ型変更のためだけに作ったマイグレーションファイルは削除した。

ham-cap commented:
修正しました！
ご確認よろしくお願いいたします。

yoshitsugu commented on db/migrate/20220601213723_add_nickname_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddNicknameToUsers < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :nickname, :string
```
同様に、PRマージ前であれば `db/migrate/20220523212231_add_zipcode_address_introduce_to_users.rb` に混ぜてしまっても大丈夫ですよ :+1:

yoshitsugu reviewed (approved):
OKです〜
